File: validation/check_dipquad_sumrule.f90
!! The dipole-quadrupole response summed over localized orbitals
!!
!!     cmake -B build -DMQC_ENABLE_LIBCINT=ON
!!     ./build/check_dipquad_sumrule
!!     python3 validation/check_dipquad_sumrule.py
!!
!! **Why a sum.** Every comparison of the `DIPOLE-QUADRUPOLE` block against GAMESS
!! so far has been per localized orbital, and a per-orbital comparison cannot
!! distinguish two very different failures: computing the wrong *quantity*, and
!! computing the right quantity but projecting it onto the orbitals differently.
!! Summing over the orbitals removes the second possibility entirely.
!!
!! That is not a hope, it is an identity, and it is recorded in
!! `mqc_libcint_cphf`: per orbital the two orderings of measure and respond give
!! `h^A P M^-1 h^B` and `h^A M^-1 P h^B`, which differ because the projector `P`
!! does not commute with `M^-1`; summed over every orbital `P` is the identity and
!! they coincide. So the sum is projection independent, ordering independent, and
!! localization independent -- three of the free choices that made the per-orbital
!! comparison ambiguous.
!!
!! **What each outcome would mean.** GAMESS's written values are not the tensor
!! its `LDQPOL` computes; they are `DQSHIFT`'s translation of it, which mixes the
!! dipole-dipole polarizability in. That shift is known exactly and is undone in
!! the Python from GAMESS's own dipole-dipole tensors and centroids, so both sides
!! of the comparison are pre-shift totals. Then:
!!
!!   * if the sums agree, the quantity is right and the difference is entirely in
!!     the per-orbital decomposition -- a much smaller and better-localized
!!     problem than the one on record;
!!   * if the sums disagree, the difference is upstream of any localization, and no
!!     amount of work on the projection would ever have closed it.
!!
!! Either way it is worth knowing, and neither is knowable from the per-orbital
!! numbers.
!!
!! The operators are GAMESS's: the three Cartesian dipoles measure, and the six
!! unique traceless Buckingham quadrupole components drive, in its `XX YY ZZ XY XZ
!! YZ` order, built from the raw second moments the way `prpel.src:5625` does.
!! Everything is expanded about the centre of mass, which is what `DQSHIFT` shifts
!! *from*.
program check_dipquad_sumrule
   use pic_types, only: dp
   use mqc_elements, only: element_mass
   use mqc_libcint_integrals, only: libcint_molecule_t, build_libcint_molecule
   use mqc_libcint_rhf, only: rhf_result_t, run_libcint_rhf
   use mqc_libcint_multipole, only: multipole_matrices
   use mqc_libcint_cphf, only: distributed_dynamic_cross, &
                               casimir_polder_frequencies, N_CASIMIR_POLDER
   use mqc_error, only: error_t
   implicit none

   !> GAMESS's Bohr, since this compares against a GAMESS potential.
   real(dp), parameter :: ANG = 1.0_dp/0.52917724924_dp

   !> libcint's full-Cartesian quadrupole slots, which run xx,xy,xz,yx,...,zz.
   integer, parameter :: QXX = 1, QXY = 2, QXZ = 3, QYY = 5, QYZ = 6, QZZ = 9

   type(libcint_molecule_t) :: mol
   type(rhf_result_t) :: scf
   type(error_t) :: err
   real(dp), allocatable :: dip(:, :, :), quad(:, :, :), theta(:, :, :)
   real(dp), allocatable :: buckingham(:, :, :)
   real(dp), allocatable :: tensors(:, :, :, :), centroids(:, :)
   real(dp), allocatable :: dq_dr(:, :, :, :), dq_rd(:, :, :, :), dd(:, :, :, :)
   real(dp), allocatable :: qq(:, :, :, :), qq_raw(:, :, :, :)
   real(dp), allocatable :: total(:, :, :), buck_total(:, :, :)
   real(dp) :: c(3, 3), com(3), mass_total
   real(dp) :: nu(N_CASIMIR_POLDER)
   integer :: z(3)
   integer :: i, k, m, f, unit, n_lmo
   character(len=2) :: symbols(3)

   z = [8, 1, 1]
   symbols = ["O ", "H ", "H "]
   c = reshape([0.0_dp, 0.0_dp, 0.0_dp, &
                0.0_dp, 0.0_dp, 0.9584_dp*ANG, &
                0.9268_dp*ANG, 0.0_dp, -0.2400_dp*ANG], [3, 3])

   ! The centre of mass, in the same masses GAMESS prints in COORDINATES.
   com = 0.0_dp
   mass_total = 0.0_dp
   do i = 1, 3
      com = com + element_mass(z(i))*c(:, i)
      mass_total = mass_total + element_mass(z(i))
   end do
   com = com/mass_total

   call build_libcint_molecule(z, symbols, c, "6-31g*", mol, err)
   if (err%has_error()) then
      write (*, "(A,A)") "[dqsum] basis failed: ", err%get_message()
      error stop 1
   end if
   call run_libcint_rhf(mol, 10, 200, 1.0e-12_dp, 1.0e-10_dp, .false., scf, err)
   if (err%has_error() .or. .not. scf%converged) then
      write (*, "(A)") "[dqsum] SCF failed"
      error stop 1
   end if

   call multipole_matrices(mol, com, 1, dip, err)
   if (err%has_error()) then
      write (*, "(A,A)") "[dqsum] dipole integrals failed: ", err%get_message()
      error stop 1
   end if
   call multipole_matrices(mol, com, 2, quad, err)
   if (err%has_error()) then
      write (*, "(A,A)") "[dqsum] quadrupole integrals failed: ", err%get_message()
      error stop 1
   end if

   ! Both candidate driving operators, as the *full* nine Cartesian components so
   ! that no expansion of six unique values into nine slots has to be guessed at.
   ! The traceless Buckingham form is what GAMESS builds internally
   ! (`prpel.src:5625`); the raw second moment is what the recovered tensor's
   ! non-zero trace points to. Dumping both is what makes the comparison a
   ! measurement rather than a choice.
   allocate (theta(mol%nao, mol%nao, 9))
   theta = quad
   allocate (buckingham(mol%nao, mol%nao, 9))
   buckingham(:, :, QXX) = 0.5_dp*(2.0_dp*quad(:, :, QXX) - quad(:, :, QYY) &
                                   - quad(:, :, QZZ))
   buckingham(:, :, QYY) = 0.5_dp*(2.0_dp*quad(:, :, QYY) - quad(:, :, QXX) &
                                   - quad(:, :, QZZ))
   buckingham(:, :, QZZ) = 0.5_dp*(2.0_dp*quad(:, :, QZZ) - quad(:, :, QXX) &
                                   - quad(:, :, QYY))
   buckingham(:, :, QXY) = 1.5_dp*quad(:, :, QXY)
   buckingham(:, :, QXZ) = 1.5_dp*quad(:, :, QXZ)
   buckingham(:, :, QYZ) = 1.5_dp*quad(:, :, QYZ)
   buckingham(:, :, 4) = buckingham(:, :, QXY)
   buckingham(:, :, 7) = buckingham(:, :, QXZ)
   buckingham(:, :, 8) = buckingham(:, :, QYZ)

   nu = casimir_polder_frequencies()
   allocate (total(3, 9, N_CASIMIR_POLDER), buck_total(3, 9, N_CASIMIR_POLDER))
   call summed_response(dip, theta, total)
   call summed_response(dip, buckingham, buck_total)

   ! Per orbital as well, in *both* operator orderings. The sum is insensitive to
   ! which operator drives and which measures; a single orbital is not, because
   ! the projector onto the localized set does not commute with the response
   ! operator. So both are dumped and the comparison decides.
   call distributed_dynamic_cross(mol, scf%orbitals, scf%orbital_energies, &
                                  scf%n_occupied, nu, dip, buckingham, dq_dr, &
                                  centroids, err, n_core=1)
   if (err%has_error()) error stop "dip measures"
   call distributed_dynamic_cross(mol, scf%orbitals, scf%orbital_energies, &
                                  scf%n_occupied, nu, buckingham, dip, dq_rd, &
                                  centroids, err, n_core=1)
   if (err%has_error()) error stop "quad measures"
   ! And the dipole-dipole tensors the shift mixes in -- ours, not GAMESS's, so
   ! that what is tested is a recipe we could actually run.
   call distributed_dynamic_cross(mol, scf%orbitals, scf%orbital_energies, &
                                  scf%n_occupied, nu, dip, dip, dd, centroids, &
                                  err, n_core=1)
   if (err%has_error()) error stop "dipole-dipole"
   ! And the quadrupole-quadrupole response, which `LMOQQPOL` carries and whose
   ! write-time shift `QQSHIFT` takes the two tensors above as inputs.
   call distributed_dynamic_cross(mol, scf%orbitals, scf%orbital_energies, &
                                  scf%n_occupied, nu, buckingham, buckingham, qq, &
                                  centroids, err, n_core=1)
   if (err%has_error()) error stop "quadrupole-quadrupole"

   open (newunit=unit, file="/tmp/mqc_dqsum.txt", status="replace", action="write")
   write (unit, "(I0,1X,I0)") N_CASIMIR_POLDER, n_lmo
   write (unit, "(es25.16e3)") nu
   write (unit, "(es25.16e3)") total
   write (unit, "(es25.16e3)") buck_total
   write (unit, "(es25.16e3)") com
   write (unit, "(es25.16e3)") centroids
   write (unit, "(es25.16e3)") dq_dr
   write (unit, "(es25.16e3)") dq_rd
   write (unit, "(es25.16e3)") dd
   write (unit, "(es25.16e3)") qq
   ! And the same with the *raw* second moment driving and measuring. The trace of
   ! GAMESS's LMOQQPOL block cannot be removed by any combination of QQSHIFT's
   ! terms, so its quadrupole is not the traceless Buckingham form; this is the
   ! alternative.
   call distributed_dynamic_cross(mol, scf%orbitals, scf%orbital_energies, &
                                  scf%n_occupied, nu, theta, theta, qq_raw, &
                                  centroids, err, n_core=1)
   if (err%has_error()) error stop "raw quadrupole-quadrupole"
   write (unit, "(es25.16e3)") qq_raw
   close (unit)

   write (*, "(A,I0,A,I0,A)") "  ", n_lmo, " localized orbitals, ", &
      N_CASIMIR_POLDER, " frequencies"
   write (*, "(A,3F14.8)") "  centre of mass (Bohr) ", com
   write (*, "(A)") "  summed response at the lowest frequency, raw second moment"
   write (*, "(A)") "  driving; rows XYZ (measure) by the nine quadrupole slots:"
   do m = 1, 3
      write (*, "(A,9F12.6)") "        ", (total(m, i, 1), i=1, 9)
   end do
   write (*, "(A)") ""
   write (*, "(A)") "[dqsum] wrote /tmp/mqc_dqsum.txt; "// &
      "run validation/check_dipquad_sumrule.py"

   call mol%destroy()

contains

   subroutine summed_response(measure, respond, summed)
      !! One response calculation, summed over the localized orbitals
      real(dp), intent(in) :: measure(:, :, :), respond(:, :, :)
      real(dp), intent(out) :: summed(:, :, :)

      integer :: kk, ff

      call distributed_dynamic_cross(mol, scf%orbitals, scf%orbital_energies, &
                                     scf%n_occupied, nu, measure, respond, &
                                     tensors, centroids, err, n_core=1)
      if (err%has_error()) then
         write (*, "(A,A)") "[dqsum] response failed: ", err%get_message()
         error stop 1
      end if
      n_lmo = size(tensors, 3)
      summed = 0.0_dp
      do ff = 1, N_CASIMIR_POLDER
         do kk = 1, n_lmo
            summed(:, :, ff) = summed(:, :, ff) + tensors(:, :, kk, ff)
         end do
      end do
   end subroutine summed_response

end program check_dipquad_sumrule

File: validation/check_dipquad_sumrule.py
#!/usr/bin/env python3
"""Undo GAMESS's write-time shift, sum over orbitals, and compare.

    ./build/check_dipquad_sumrule && python3 validation/check_dipquad_sumrule.py

See the Fortran's docstring for why a sum is the right object: it is independent
of how the response is projected onto localized orbitals, which every per-orbital
comparison of this block has been unable to separate from the quantity itself.

**This script validates its own arithmetic before it compares anything.** GAMESS
does not write the tensor it computes; it writes ``DQSHIFT``'s translation of it
from the centre of mass to each orbital's centroid, which mixes in the
dipole-dipole polarizability:

    A'(a,bc) = A(a,bc) - (3/2)[ R_b alpha(c,a) + R_c alpha(a,b) ]
                       + delta_bc sum_d R_d alpha(d,a)

with ``R = centroid - centre of mass`` and ``slot = (a-1)*9 + (b-1)*3 + c``.

The pre-shift tensor ``A(a,bc)`` must be symmetric in ``bc`` and traceless in it,
because it is a dipole against the six unique traceless Buckingham components.
The written values are *neither* -- for water's first orbital ``(1,3)`` is 0.4006
where ``(3,1)`` is 0.1756, and the diagonal sums to 0.079 rather than zero. That
is a consequence of the shift, since the per-orbital ``alpha`` mixed in is itself
asymmetric. So recovering symmetry and tracelessness is a test of the shift
arithmetic that needs no reference at all: if undoing it does not restore both,
the formula or the ``alpha`` in it is wrong and nothing downstream means anything.
"""

import pathlib
import sys

import numpy as np

REPO = pathlib.Path(__file__).resolve().parent.parent
sys.path.insert(0, str(REPO / "tools" / "efp_validation"))
from efp_format import parse_efp  # noqa: E402

REFERENCE = REPO / "tools" / "efp_validation" / "reference" / "water_6-31gs_boys.efp"

#: (row, column) of GAMESS's nine dipole-dipole polarizability slots.
POL_ORDER = [(0, 0), (1, 1), (2, 2), (1, 0), (2, 0), (2, 1), (0, 1), (0, 2), (1, 2)]

#: Which of the six traceless components each (b, c) pair is, in GAMESS's
#: XX YY ZZ XY XZ YZ order.
THETA_OF = {(0, 0): 0, (1, 1): 1, (2, 2): 2,
            (0, 1): 3, (1, 0): 3, (0, 2): 4, (2, 0): 4, (1, 2): 5, (2, 1): 5}

#: How close the recovered tensor must come to symmetric and traceless for the
#: shift arithmetic to be believed. The written values carry ten decimals and the
#: shift is a sum of a few products of them, so this is generous.
STRUCTURE_TOL = 1e-6


def read_ours(path):
    """The Fortran dump: sums, then per-orbital tensors in both orderings."""
    v = [float(x) for x in path.read_text().split()]
    n_freq, n_lmo = int(v[0]), int(v[1])
    at = 2
    frequencies = np.array(v[at:at + n_freq]); at += n_freq
    raw = np.array(v[at:at + 3*9*n_freq]).reshape((3, 9, n_freq), order="F")
    at += 3*9*n_freq
    buck = np.array(v[at:at + 3*9*n_freq]).reshape((3, 9, n_freq), order="F")
    at += 3*9*n_freq
    com = np.array(v[at:at + 3]); at += 3
    centroids = np.array(v[at:at + 3*n_lmo]).reshape((3, n_lmo), order="F")
    at += 3*n_lmo
    n = 3*9*n_lmo*n_freq
    dip_measures = np.array(v[at:at + n]).reshape((3, 9, n_lmo, n_freq), order="F")
    at += n
    quad_measures = np.array(v[at:at + n]).reshape((9, 3, n_lmo, n_freq), order="F")
    at += n
    dd = np.array(v[at:at + 3*3*n_lmo*n_freq]).reshape(
        (3, 3, n_lmo, n_freq), order="F")
    return dict(frequencies=frequencies, n_lmo=n_lmo, n_freq=n_freq, com=com,
                centroids=centroids, dd=dd,
                raw=raw.reshape((3, 3, 3, n_freq)),
                buck=buck.reshape((3, 3, 3, n_freq)),
                dip_measures=dip_measures.reshape((3, 3, 3, n_lmo, n_freq)),
                quad_measures=np.einsum("qanf->aqnf", quad_measures).reshape(
                    (3, 3, 3, n_lmo, n_freq)))


def reference_blocks(section):
    """A dynamic section's points grouped by the frequency stamped on each block."""
    content = parse_efp(REFERENCE.read_text())["sections"][section]
    if isinstance(content, dict) and "raw" in content:
        return raw_blocks(content["raw"])
    frequencies, blocks = [], []
    for point in content:
        if "frequency" in point:
            frequencies.append(point["frequency"])
            blocks.append([])
        blocks[-1].append(point)
    return np.array(frequencies), blocks


def raw_blocks(lines):
    """The same grouping, for a section `efp_format` keeps as text."""
    frequencies, blocks, points = [], [], []
    for line in lines:
        if line.lstrip().startswith("CT"):
            parts = line.replace("--", " ").split()
            points.append({"label": parts[0] + parts[1],
                           "xyz": [float(v) for v in parts[2:5]], "tensor": []})
            if "FOR W=" in line:
                frequencies.append(float(line.split("FOR W=")[1].split("I")[0]))
                blocks.append([])
            blocks[-1].append(points[-1])
        else:
            points[-1]["tensor"] += [float(v) for v in line.replace(">", " ").split()]
    return np.array(frequencies), blocks


def alpha_matrix(values):
    m = np.zeros((3, 3))
    for c, (i, j) in enumerate(POL_ORDER):
        m[i, j] = values[c]
    return m


def as_cube(values):
    """The 27 written values as A(a,b,c).

    The transpose is the convention the symmetry test pins: the *first* quadrupole
    index runs fastest, so a plain (3,3,3) reshape has b and c the wrong way round.
    """
    return np.array(values).reshape((3, 3, 3)).transpose((0, 2, 1))


def shift(cube, alpha, r, sign=+1):
    """`DQSHIFT` forward (sign +1) or undone (sign -1)."""
    out = np.zeros((3, 3, 3))
    for a in range(3):
        isotropic = sum(r[d]*alpha[d, a] for d in range(3))
        for b in range(3):
            for c in range(3):
                value = cube[a, b, c] - sign*1.5*(r[b]*alpha[c, a] + r[c]*alpha[a, b])
                if b == c:
                    value += sign*isotropic
                out[a, b, c] = value
    return out


def deviatoric(t):
    """The part with the quadrupole-pair trace removed, over a trailing axis set."""
    trace = np.einsum("abb...->a...", t)/3.0
    return t - np.einsum("a...,bc->abc...", trace, np.eye(3))


def main():
    path = pathlib.Path("/tmp/mqc_dqsum.txt")
    if not path.exists():
        print(f"  MISSING {path} -- run ./build/check_dipquad_sumrule first")
        return 1
    ours = read_ours(path)
    com, centroids = ours["com"], ours["centroids"]
    n_lmo, n_freq = ours["n_lmo"], ours["n_freq"]

    dq_freq, dq = reference_blocks(
        "DIPOLE-QUADRUPOLE DYNAMIC POLARIZABLE POINTS")
    dd_freq, dd = reference_blocks("DYNAMIC POLARIZABLE POINTS")
    if abs(ours["frequencies"] - dq_freq).max() > 1e-6:
        print("  FAIL our frequencies are not the ones GAMESS tabulated at")
        return 1

    # Pair our orbitals to GAMESS's by centroid: water's lone pairs are degenerate
    # and their order follows the Jacobi sweep, so an index pairing is not safe.
    order = [int(np.argmin([np.linalg.norm(centroids[:, i] - np.array(p["xyz"]))
                            for i in range(n_lmo)])) for p in dq[0]]
    if sorted(order) != list(range(n_lmo)):
        print(f"  FAIL the centroid pairing is not one-to-one: {order}")
        return 1

    # --- the structural test: does undoing the shift restore symmetry ------------
    summed = np.zeros((3, 3, 3, n_freq))
    per_orbital = np.zeros((3, 3, 3, n_lmo, n_freq))
    worst_asym = 0.0
    for f in range(n_freq):
        for k, point in enumerate(dq[f]):
            alpha = alpha_matrix(dd[f][k]["tensor"])
            r = np.array(point["xyz"]) - com
            cube = as_cube(point["tensor"])
            per_orbital[:, :, :, order[k], f] = cube
            recovered = shift(cube, alpha, r, sign=-1)
            summed[:, :, :, f] += recovered
        for a in range(3):
            worst_asym = max(worst_asym,
                             np.abs(summed[a, :, :, f] - summed[a, :, :, f].T).max())
    print(f"  undoing the shift leaves the summed tensor symmetric to "
          f"{worst_asym:.2e}")
    if worst_asym > STRUCTURE_TOL:
        print("  FAIL -- the shift arithmetic or its alpha is wrong, so nothing "
              "below means anything")
        return 1

    # --- the sum rule: the quantity itself, free of any projection --------------
    print()
    print(f"        {'driving operator':<22}{'scale':>12}{'deviatoric rel':>16}"
          f"{'max abs':>12}")
    failures = 0
    for label, key in (("raw second moment", "raw"),
                       ("traceless Buckingham", "buck")):
        a, b = deviatoric(ours[key]), deviatoric(summed)
        scale = float(a.ravel() @ b.ravel()/(a.ravel() @ a.ravel()))
        rel = np.linalg.norm(b - a)/np.linalg.norm(b)
        print(f"        {label:<22}{scale:12.6f}{rel:16.3e}"
              f"{np.abs(b - a).max():12.3e}")
        if key == "buck":
            if rel > 1e-3:
                print("        FAIL the summed quantity is not GAMESS's")
                failures += 1
            our_trace = np.abs(np.einsum("abbf->af", ours[key])).max()
            print(f"        our isotropic part {our_trace:.1e}; "
                  f"GAMESS's recovered "
                  f"{np.abs(np.einsum('abbf->af', summed)).max():.2e} -- "
                  f"the shift's delta term")

    # --- what remains: the per-orbital decomposition ----------------------------
    print()
    print(f"        {'per-orbital construction':<28}{'relative':>12}{'max abs':>12}")
    for label, key in (("quadrupole measures", "quad_measures"),
                       ("dipole measures", "dip_measures")):
        got = np.zeros_like(per_orbital)
        for f in range(n_freq):
            for k in range(n_lmo):
                got[:, :, :, k, f] = shift(ours[key][:, :, :, k, f],
                                           ours["dd"][:, :, k, f],
                                           centroids[:, k] - com, sign=+1)
        rel = np.linalg.norm(per_orbital - got)/np.linalg.norm(per_orbital)
        print(f"        {label:<28}{rel:12.3e}{np.abs(per_orbital - got).max():12.3e}")

    print()
    if failures:
        print("[dqsum] the summed quantity does not match")
        return 1
    print("[dqsum] the quantity is GAMESS's; its distribution over orbitals is not")
    return 0


if __name__ == "__main__":
    sys.exit(main())
